### PR TITLE
#779 Last space trim when typing fast

### PR DESCRIPTION
**Pull request by FreeFrags**
optional st-trim-search-input added to the table if set it will use the set value if not set it will use false

### CODE REVIEW

File: test/spec/stTable.spec.js [hunk 1]
```diff
@@ -177,14 +177,14 @@ describe('st table Controller', function () {
...
-        ]);
-      });
+      //it('should trim if the input is a string', function () {
+      //  ctrl.search(' re ', 'name');
+      //  expect(scope.data).toEqual([
+      //    {name: 'Renard', firstname: 'Laurent', age: 66},
+      //    {name: 'Renard', firstname: 'Olivier', age: 33},
+      //    {name: 'Faivre', firstname: 'Blandine', age: 44}
+      //  ]);
+      //});
```
Comment (josh-freislich) [inline test/spec/stTable.spec.js]: Hi @FreeFrags, could you update the tests for trimSearchInput = true|false?

**Comment by FreeFrags** [comment]
@josh-freislich updated the test as requested

File: test/spec/stTable.spec.js [hunk 2]
```diff
@@ -177,7 +178,7 @@ describe('st table Controller', function () {
         ]);
       });
 
-      it('should trim if the input is a string', function () {
+      it('should trim if the input is a string when trimSearchInput is set to true', function () {
```
Comment (josh-freislich) [inline test/spec/stTable.spec.js]: test for false?

**Comment by FreeFrags** [comment]
@josh-freislich as requested

File: src/stTable.js [hunk 3]
```diff
@@ -113,7 +119,10 @@ ng.module('smart-table').controller('stTableController', [
       var predicateObject = tableState.search.predicateObject || {};
       var prop = predicate ? predicate : '$';
 
-      input = ng.isString(input) ? input.trim() : input;
+      if (trimSearchInput && ng.isString(input)) {
+        input = input.trim();
+      }
```
Comment (josh-freislich) [inline src/stTable.js]: else?

File: test/spec/stTable.spec.js [hunk 4]
```diff
@@ -22,7 +22,7 @@ describe('st table Controller', function () {
       scope.data = dataSet;
       ctrl = $controller('stTableController', {
         $scope: scope, $parse: $parse, $filter: $filter, $attrs: {
-          stTable: 'data'
+          stTable: 'data',
```
Comment (josh-freislich) [inline test/spec/stTable.spec.js]: comma not necessary.

File: test/spec/stTable.spec.js [hunk 5]
```diff
@@ -380,6 +371,109 @@ describe('st table Controller', function () {
...
+        ctrl.search(' ', 'name');
+        expect(scope.data).toEqual([
+          {name: ' Starting with a space'},
+          {name: 'Containing a space'},
+          {name: 'Ending with a space '}
+        ]);
+      });
+ 
+      it('should NOT trim and search for everything containing the search string including space and s', function () {
+        ctrl.search(' S', 'name');
```
Comment (josh-freislich) [inline test/spec/stTable.spec.js]: Good tests. It captures behaviour on space, leading space, trailing space. 

Consider keeping "s|S" consistent to help those who are not that strong in javascript to not be distracted by the change in casing.

**Comment by lorenzofox3** [comment]
Hi, and thank you everybody for the work and review here. I would rather see this option (and the trimming ) at the level of stSearchDirective rather than on the table controller as it is not global to all directives. I believe it will be more consistent. 

That would imply: 
1. passing a flag or taking the default value from the config (https://github.com/lorenzofox3/Smart-Table/blob/master/src/stSearch.js#L9)
2. trimming the input https://github.com/lorenzofox3/Smart-Table/blob/master/src/stSearch.js#L15 if flag is on **and** input is a string.
3. set the default value on the global config https://github.com/lorenzofox3/Smart-Table/blob/master/src/stConfig.js (false by default would be better)

for the tests I suggest on test case when flag is on and input is a string and another one when flag is on and input is not a string. 

I'll close that PR so you can start with a fresh start. 

If you want me to do it, I don't mind (I have a bit more time now) but I think that would make a good first contribution for you as you have already passed some time doing it.

Let me know if you need any help.

Cheers

**Pull request closed without merging**